scripts/TPI.py

```diff
@@ -0,0 +1,50 @@
+import numpy as np
+import scipy.optimize as opt
+import firm
+import households as hh
+import aggregates as agg
+
+
+def solve_tp(r_path_init, BQ_path_init, params):
+    '''
+    Solves for the time path equlibrium using TPI
+    '''
+    (beta, sigma, alpha, A, delta, theta, T, xi, b_sp1_pre, r_ss, BQ_ss, b_sp1_ss) = params
```

Comment:
What is theta?

Reply:
@jdebacker  Elasticity of labor supply to be used in `FOCs` based on what is defined as arguments in `households.py`

Reply:
OK - we're going to use the elliptical utility function here, which is defined by the parameters `b_ellip` and `upsilon`.

Reply:
@jdebacker I noticed Leili and Milton are working on this but I just ask this for my information and understanding. What if I solve like this:

`# solve upper right elements before the first full lifetime
        UpMaskb = np.triu(np.ones((T + S, S)), 1) #I replaced -1 with 1 to keep the diagonal equal to zero
        for t in range(T + S):
            foc_args = (beta, sigma, r_path[t:t+S], w_path[t:t+S], b_ellip, upsilon, chi, rho_s)
            b_sp1_guess = b_sp1_ss
            result = opt.root(hh.FOCs, b_sp1_guess, args=foc_args)
            b_sp1_mat[:, t:t+S] = (UpMaskb * result.x + b_sp1_mat[:, t:t+S])
            euler_errors_mat[:, t:t+S] = (UpMaskb * result.fun + euler_errors_mat[:, t:t+S])`

I changed the indices for time path and as my `UpMaskb` is an upper triangular matrix, all elements at and below the diagonal -- which are full time path decision makers -- will be zero after multiplication to results.

Reply:
@saharnazb First, I thought we had `b_sp1_mat` as a  TxS matrix.  Second, I'm not 100% sure, but I don't think this will work. I think as you loop over the t's you are going to write over some results that you don't want to write over, since the `UpMaskb` is non-zero for all elements above the diagonal -- and as you lay the solution in the "twist donut", you are only working along the diagonal.

But you could test this by setting up a simple example in a script of notebook and seeing if it works as expected.